<a href="https://colab.research.google.com/github/MiracleCakra/Machine-Learning_Ganjil_2025/blob/main/Week13_JS13_Artificial%20Neural%20Network%20(ANN)%20dan%20Evaluasi%20Classifier/Tugas_Praktikum_JS13_Artificial_Neural_Network_(ANN)_dan_Evaluasi_Classifier.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Tugas Praktikum
Gunakan JST untuk klasifikasi angka tulisan tangan (MNIST).

Langkah:

Load dataset MNIST dari Keras.

Bangun model dengan 2 hidden layer.

Latih model dan evaluasi akurasi.

In [1]:
import tensorflow as tf
import time # Import library time untuk mengukur durasi pelatihan
from tensorflow.keras.datasets import mnist
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Flatten
from tensorflow.keras.utils import to_categorical

# --- 1. Load & Preprocess Data ---
(X_train, y_train), (X_test, y_test) = mnist.load_data()

# Normalisasi data (0-255 -> 0-1)
X_train = X_train / 255.0
X_test = X_test / 255.0

# One-hot encoding label
y_train = to_categorical(y_train, 10)
y_test = to_categorical(y_test, 10)

# --- 2. Bangun Model JST ---
model = Sequential([
    # Input Layer: Ubah gambar 28x28 menjadi vektor 1D (784 fitur)
    Flatten(input_shape=(28, 28)),

    # Hidden Layer 1: 128 neuron, aktivasi ReLU
    Dense(128, activation='relu'),

    # Hidden Layer 2: 64 neuron, aktivasi ReLU
    Dense(64, activation='relu'),

    # Output Layer: 10 kelas (angka 0-9), aktivasi Softmax (wajib untuk klasifikasi multikelas)
    Dense(10, activation='softmax')
])

# --- 3. Kompilasi Model ---
model.compile(optimizer='adam',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# --- 4. Latih Model ---
print("Mulai pelatihan...")
start_time = time.time() # Mulai stopwatch

# Epochs=10 biasanya cukup untuk MNIST mencapai akurasi >97%
history = model.fit(X_train, y_train, epochs=10, batch_size=32, validation_split=0.2)

end_time = time.time() # Hentikan stopwatch
print(f"Waktu pelatihan: {end_time - start_time:.2f} detik")

# --- 5. Evaluasi Model ---
print("\nEvaluasi pada data uji:")
loss, acc = model.evaluate(X_test, y_test)
print(f"Akurasi Final: {acc:.4f}")

11490434/11490434 ━━━━━━━━━━━━━━━━━━━━ 0s 0us/step


/usr/local/lib/python3.12/dist-packages/keras/src/layers/reshaping/flatten.py:37: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)


Mulai pelatihan...
Epoch 1/10
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 9s 5ms/step - accuracy: 0.8625 - loss: 0.4742 - val_accuracy: 0.9602 - val_loss: 0.1353
Epoch 2/10
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 8s 5ms/step - accuracy: 0.9673 - loss: 0.1093 - val_accuracy: 0.9694 - val_loss: 0.1008
Epoch 3/10
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 7s 4ms/step - accuracy: 0.9780 - loss: 0.0736 - val_accuracy: 0.9675 - val_loss: 0.1078
Epoch 4/10
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 8s 5ms/step - accuracy: 0.9829 - loss: 0.0554 - val_accuracy: 0.9712 - val_loss: 0.1015
Epoch 5/10
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 7s 5ms/step - accuracy: 0.9862 - loss: 0.0429 - val_accuracy: 0.9761 - val_loss: 0.0852
Epoch 6/10
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 7s 5ms/step - accuracy: 0.9890 - loss: 0.0323 - val_accuracy: 0.9703 - val_loss: 0.1092
Epoch 7/10
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 8s 5ms/step - accuracy: 0.9914 - loss: 0.0256 - val_accuracy: 0.9724 - val_loss: 0.1053
Epoch 8/10
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 7s 4ms/step - accuracy: 0.992

## A. Ubah Jumlah Neuron (Misal: 256 dan 128)

In [2]:
import tensorflow as tf
import time # Import library time untuk mengukur durasi pelatihan
from tensorflow.keras.datasets import mnist
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Flatten
from tensorflow.keras.utils import to_categorical

# --- 1. Load & Preprocess Data ---
(X_train, y_train), (X_test, y_test) = mnist.load_data()

# Normalisasi data (0-255 -> 0-1)
X_train = X_train / 255.0
X_test = X_test / 255.0

# One-hot encoding label
y_train = to_categorical(y_train, 10)
y_test = to_categorical(y_test, 10)

# --- 2. Bangun Model JST ---
model = Sequential([
    Flatten(input_shape=(28, 28)),
    Dense(256, activation='relu'), # Diubah dari 128 ke 256
    Dense(128, activation='relu'), # Diubah dari 64 ke 128
    Dense(10, activation='softmax')
])

# --- 3. Kompilasi Model ---
model.compile(optimizer='adam',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# --- 4. Latih Model ---
print("Mulai pelatihan...")
start_time = time.time() # Mulai stopwatch

# Epochs=10 biasanya cukup untuk MNIST mencapai akurasi >97%
history = model.fit(X_train, y_train, epochs=10, batch_size=32, validation_split=0.2)

end_time = time.time() # Hentikan stopwatch
print(f"Waktu pelatihan: {end_time - start_time:.2f} detik")

# --- 5. Evaluasi Model ---
print("\nEvaluasi pada data uji:")
loss, acc = model.evaluate(X_test, y_test)
print(f"Akurasi Final: {acc:.4f}")

Mulai pelatihan...
Epoch 1/10
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 11s 6ms/step - accuracy: 0.8844 - loss: 0.3951 - val_accuracy: 0.9588 - val_loss: 0.1374
Epoch 2/10
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 11s 7ms/step - accuracy: 0.9697 - loss: 0.0939 - val_accuracy: 0.9713 - val_loss: 0.0917
Epoch 3/10
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 10s 7ms/step - accuracy: 0.9816 - loss: 0.0609 - val_accuracy: 0.9692 - val_loss: 0.1047
Epoch 4/10
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 10s 7ms/step - accuracy: 0.9867 - loss: 0.0434 - val_accuracy: 0.9728 - val_loss: 0.1045
Epoch 5/10
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 11s 7ms/step - accuracy: 0.9877 - loss: 0.0364 - val_accuracy: 0.9751 - val_loss: 0.0928
Epoch 6/10
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 9s 6ms/step - accuracy: 0.9913 - loss: 0.0263 - val_accuracy: 0.9729 - val_loss: 0.0950
Epoch 7/10
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 10s 7ms/step - accuracy: 0.9926 - loss: 0.0221 - val_accuracy: 0.9726 - val_loss: 0.1178
Epoch 8/10
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 10s 7ms/step - accuracy

## B. Tambahkan Satu Hidden Layer Lagi

In [3]:
import tensorflow as tf
import time # Import library time untuk mengukur durasi pelatihan
from tensorflow.keras.datasets import mnist
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Flatten
from tensorflow.keras.utils import to_categorical

# --- 1. Load & Preprocess Data ---
(X_train, y_train), (X_test, y_test) = mnist.load_data()

# Normalisasi data (0-255 -> 0-1)
X_train = X_train / 255.0
X_test = X_test / 255.0

# One-hot encoding label
y_train = to_categorical(y_train, 10)
y_test = to_categorical(y_test, 10)

# --- 2. Bangun Model JST ---
model = Sequential([
    Flatten(input_shape=(28, 28)),
    Dense(128, activation='relu'),
    Dense(64, activation='relu'),
    Dense(32, activation='relu'), # Layer tambahan
    Dense(10, activation='softmax')
])

# --- 3. Kompilasi Model ---
model.compile(optimizer='adam',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# --- 4. Latih Model ---
print("Mulai pelatihan...")
start_time = time.time() # Mulai stopwatch

# Epochs=10 biasanya cukup untuk MNIST mencapai akurasi >97%
history = model.fit(X_train, y_train, epochs=10, batch_size=32, validation_split=0.2)

end_time = time.time() # Hentikan stopwatch
print(f"Waktu pelatihan: {end_time - start_time:.2f} detik")

# --- 5. Evaluasi Model ---
print("\nEvaluasi pada data uji:")
loss, acc = model.evaluate(X_test, y_test)
print(f"Akurasi Final: {acc:.4f}")

Mulai pelatihan...
Epoch 1/10
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 10s 6ms/step - accuracy: 0.8507 - loss: 0.5022 - val_accuracy: 0.9587 - val_loss: 0.1447
Epoch 2/10
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 8s 5ms/step - accuracy: 0.9632 - loss: 0.1231 - val_accuracy: 0.9669 - val_loss: 0.1118
Epoch 3/10
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 8s 5ms/step - accuracy: 0.9745 - loss: 0.0834 - val_accuracy: 0.9691 - val_loss: 0.1013
Epoch 4/10
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 8s 5ms/step - accuracy: 0.9812 - loss: 0.0609 - val_accuracy: 0.9640 - val_loss: 0.1267
Epoch 5/10
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 7s 5ms/step - accuracy: 0.9853 - loss: 0.0465 - val_accuracy: 0.9744 - val_loss: 0.0962
Epoch 6/10
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 11s 5ms/step - accuracy: 0.9887 - loss: 0.0374 - val_accuracy: 0.9716 - val_loss: 0.1062
Epoch 7/10
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 8s 5ms/step - accuracy: 0.9896 - loss: 0.0309 - val_accuracy: 0.9717 - val_loss: 0.1053
Epoch 8/10
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 7s 5ms/step - accuracy: 0.9

## C. Eksperimen Aktivasi: Sigmoid vs ReLU

In [6]:
import tensorflow as tf
import time
from tensorflow.keras.datasets import mnist
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Flatten
from tensorflow.keras.utils import to_categorical

# --- 1. Load & Preprocess Data ---
(X_train, y_train), (X_test, y_test) = mnist.load_data()

# Normalisasi data (0-255 -> 0-1)
X_train = X_train / 255.0
X_test = X_test / 255.0

# One-hot encoding label
y_train = to_categorical(y_train, 10)
y_test = to_categorical(y_test, 10)

# --- 2. Bangun Model JST ---
model = Sequential([
    Flatten(input_shape=(28, 28)),
    Dense(128, activation='sigmoid'),
    Dense(64, activation='sigmoid'),
    Dense(10, activation='softmax')
])

# --- 3. Kompilasi Model ---
model.compile(optimizer='adam',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# --- 4. Latih Model ---
print("Mulai pelatihan...")
start_time = time.time() # Mulai stopwatch

history = model.fit(X_train, y_train, epochs=10, batch_size=32, validation_split=0.2)

end_time = time.time() # Hentikan stopwatch
print(f"Waktu pelatihan: {end_time - start_time:.2f} detik")

# --- 5. Evaluasi Model ---
print("\nEvaluasi pada data uji:")
loss, acc = model.evaluate(X_test, y_test)
print(f"Akurasi Final: {acc:.4f}")

Mulai pelatihan...
Epoch 1/10
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 9s 5ms/step - accuracy: 0.7636 - loss: 0.9401 - val_accuracy: 0.9336 - val_loss: 0.2325
Epoch 2/10
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 8s 5ms/step - accuracy: 0.9360 - loss: 0.2187 - val_accuracy: 0.9527 - val_loss: 0.1658
Epoch 3/10
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 10s 5ms/step - accuracy: 0.9563 - loss: 0.1478 - val_accuracy: 0.9607 - val_loss: 0.1375
Epoch 4/10
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 10s 5ms/step - accuracy: 0.9697 - loss: 0.1059 - val_accuracy: 0.9670 - val_loss: 0.1129
Epoch 5/10
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 8s 5ms/step - accuracy: 0.9753 - loss: 0.0871 - val_accuracy: 0.9705 - val_loss: 0.1010
Epoch 6/10
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 8s 5ms/step - accuracy: 0.9806 - loss: 0.0664 - val_accuracy: 0.9686 - val_loss: 0.0989
Epoch 7/10
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 7s 5ms/step - accuracy: 0.9851 - loss: 0.0528 - val_accuracy: 0.9733 - val_loss: 0.0894
Epoch 8/10
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 8s 5ms/step - accuracy: 0.9

* **Kode A (Large Neurons 256 & 128, ReLU):** Hasil 97.82% (102.67s)
* **Kode B (Deep Layers 3 Layer, ReLU):** Hasil 97.29% (86.01s)
* **Kode C (Sigmoid 128 & 64):** Hasil 97.69% (82.91s)

---

# Laporan Analisis Praktikum: Variasi Arsitektur JST pada MNIST

### 1. Ringkasan Hasil Eksperimen

| Skenario | Konfigurasi Model | Waktu (Detik) | Akurasi Final | Loss Final |
| :--- | :--- | :--- | :--- | :--- |
| **A** | **Large Neurons** (256, 128) - ReLU | 102.67 s | **97.82%** (Highest) | 0.0949 |
| **B** | **Deep Layers** (128, 64, 32) - ReLU | 86.01 s | 97.29% (Lowest) | 0.1467 |
| **C** | **Sigmoid** (128, 64) - Sigmoid | **82.91 s** (Fastest) | 97.69% | **0.0857** (Best) |

---

### 2. Analisis Mendalam

#### **A. Analisis Akurasi & Model Terbaik (Skenario A)**
* **Hasil:** Skenario A (Kode A) menjadi pemenang akurasi dengan **97.82%**.
* **Penyebab Teknis:** Kode A menggunakan jumlah neuron yang paling besar (**256** di layer pertama dan **128** di layer kedua).
* **Penjelasan:** Semakin banyak neuron, semakin banyak "kapasitas memori" (parameter) yang dimiliki jaringan untuk menangkap detail pola tulisan tangan yang rumit. Matriks bobot yang besar ini memungkinkan model membedakan variasi angka dengan lebih presisi dibandingkan model yang lebih kecil.

#### **B. Analisis "Deep Learning" yang Kurang Optimal (Skenario B)**
* **Hasil:** Skenario B (Kode B) justru memiliki akurasi terendah (**97.29%**) dan Loss tertinggi (**0.1467**), meskipun memiliki 3 hidden layer.
* **Penyebab Teknis:** Menambah kedalaman (*depth*) pada data sederhana seperti MNIST tidak selalu menjamin hasil lebih baik.
* **Penjelasan:** Kemungkinan terjadi **Overfitting** atau degradasi sinyal. Dengan arsitektur yang lebih dalam (128->64->32), model mungkin menjadi terlalu kompleks untuk data yang sebenarnya sederhana, sehingga "menghafal" noise pada data latih dan kesulitan saat diuji (terlihat dari *Loss* validasi yang tinggi: 0.14).

#### **C. Fenomena Unik Fungsi Aktivasi Sigmoid (Skenario C)**
* **Hasil:** Skenario C (Kode C) menempati posisi kedua dalam akurasi (**97.69%**) namun memiliki **Loss terendah (0.0857)**.
* **Analisis "Slow Start":**
    * Perhatikan Epoch 1 pada hasil C (Gambar 3ce7e0): Akurasi hanya **76.36%** dan Loss **0.94**. Ini jauh lebih buruk dibanding ReLU (A dan B) yang langsung start di atas 85%.
    * Ini disebabkan oleh karakteristik **Sigmoid** yang rentan terhadap *Vanishing Gradient* di awal, membuat pembelajaran lambat.
* **Analisis "Strong Finish":**
    * Meskipun lambat di awal, model ini berhasil mengejar ketertinggalan dan menghasilkan prediksi yang sangat stabil (*Loss* kecil) di akhir. Ini menunjukkan bahwa untuk jaringan yang tidak terlalu dalam (hanya 2 layer), Sigmoid masih cukup kompetitif.

#### **D. Analisis Efisiensi Waktu (Computational Cost)**
1.  **Terlambat (Skenario A - 102.67s):**
    Ini wajar karena jumlah parameter yang harus dihitung sangat masif. Layer pertama saja (784 input ke 256 neuron) memiliki bobot 2x lipat lebih banyak dibanding skenario B atau C.
2.  **Tercepat (Skenario C - 82.91s):**
    Skenario C paling cepat karena strukturnya sederhana (hanya 2 layer: 128, 64) dibandingkan B yang punya 3 layer, dan jumlah neuronnya lebih sedikit dibandingkan A.

---